seqfold/fold.go

```diff
@@ -0,0 +1,875 @@
+		if en, ok := fc.Energies.DE[pair_de]; ok {
+			d_h, d_s := en.H, en.S
+			dG += DG(d_h, d_s, fc.T)
+			return dG
+		}
+	}
+	return 0
+}
+
+// Hairpin calculates the free energy of a hairpin.
+// Args:
+//	i: The index of start of hairpin
+//	j: The index of end of hairpin
+//	fc: the folding context
+// Returns:
+//    float: The free energy increment from the hairpin structure
+func Hairpin(i, j int, fc FoldContext) float64 {
+	if j-i < 4 {
+		return math.Inf(1)
+	}
+
+	hairpinSeq := fc.Seq[i : j+1]
+	hairpin_len := len(hairpinSeq) - 2
+	pair := Pair(fc.Seq, i, i+1, j, j-1)
+
+	if fc.Energies.COMPLEMENT[hairpinSeq[0]] != hairpinSeq[len(hairpinSeq)-1] {
+		// not known terminal pair, nothing to close "hairpin"
+
+		// TODO errors
+		// raise RuntimeError
+		panic("emap.COMPLEMENT[hairpin[0]] != hairpin[-1]")
+	}
+
+	dG := 0.0
+	if fc.Energies.TRI_TETRA_LOOPS != nil {
+		if en, ok := fc.Energies.TRI_TETRA_LOOPS[hairpinSeq]; ok {
+			// it's a pre-known hairpin with known value
+			d_h, d_s := en.H, en.S
+			dG = DG(d_h, d_s, fc.T)
+		}
+	}
+
+	// add penalty based on size
+	if en, ok := fc.Energies.HAIRPIN_LOOPS[hairpin_len]; ok {
+		d_h, d_s := en.H, en.S
+		dG += DG(d_h, d_s, fc.T)
+	} else {
+		// it's too large, extrapolate
+		en := fc.Energies.HAIRPIN_LOOPS[30]
+		d_h, d_s := en.H, en.S
+		d_g_inc := DG(d_h, d_s, fc.T)
+		dG += JS(hairpin_len, 30, d_g_inc, fc.T)
+	}
+
+	// add penalty for a terminal mismatch
+	en, ok := fc.Energies.TERMINAL_MM[pair]
+	if hairpin_len > 3 && ok {
+		d_h, d_s := en.H, en.S
+		dG += DG(d_h, d_s, fc.T)
+	}
+
+	// add penalty if length 3 and AT closing, formula 8 from SantaLucia, 2004
+	if hairpin_len == 3 && (hairpinSeq[0] == 'A' || hairpinSeq[len(hairpinSeq)-1] == 'A') {
+		dG += 0.5 // convert to entropy
+	}
+
+	return dG
+}
+
+// Find the free energy given delta h, s and temp
+// Args:
+//    d_h: The enthalpy increment in kcal / mol
+//    d_s: The entropy increment in cal / mol
+//    temp: The temperature in Kelvin
+// Returns:
+//    The free energy increment in kcal / (mol x K)
+func DG(d_h, d_s, temp float64) float64 {
+	return d_h - temp*(d_s/1000.0)
+}
+
+// Estimate the free energy of length query_len based on one of length known_len.
+//
+// The Jacobson-Stockmayer entry extrapolation formula is used
+// for bulges, hairpins, etc that fall outside the 30nt upper limit
+// for pre-calculated free-energies. See SantaLucia and Hicks (2004).
+// Args:
+//     query_len: Length of element without known free energy value
+//     known_len: Length of element with known free energy value (d_g_x)
+//     d_g_x: The free energy of the element known_len
+//     temp: Temperature in Kelvin
+// Returns:
+//     float: The free energy for a structure of length query_len
+func JS(query_len, known_len int, d_g_x, temp float64) float64 {
+	gas_constant := 1.9872e-3
+	return d_g_x + 2.44*gas_constant*temp*math.Log(float64(query_len)/float64(known_len))
```

Comment:
expand this to be more commented and readable

Reply:
Renamed it with a better name, I don't really know what this is doing 😅